# **Block 1: Import only “seaborn”, “pandas”, “numpy”, “matplotlib.pyplot”, and “from sklearn.linear_model import LinearRegression”. Import the data here as well using seaborn (data = sns.load_dataset('geyser'))**

In [1]:
import seaborn as sns
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
from sklearn.linear_model import LinearRegression

data = sns.load_dataset('geyser')

# **Block 2 (with EXTRA CREDIT function): Create your own k-fold cross validation code with the number of folds being 10.**

In [5]:
def kfold_score(x,y,folds):
  
  training_r2 = []
  testing_r2 = []

  # randomize the order of the data
  data.sample(frac=1)

  X = np.array(x) 
  X = X.reshape(len(X),1)
  y = np.array(y) 

  for i in range(0,folds):  # number of folds is number of times you go through for loop

      # indexes for making the folds; divide total length of data by 10 folds
      j = int(i* (len(X)/folds)) 
      f = int(j + (len(X)/folds))
      
      # index into X and y for testing and training based on the folds
      Xtest = X[j:f]

      # turn Xtrain into an array
      Xtrain1 = X[0:j] 
      Xtrain2 = X[f:]
      Xtrain = np.concatenate((Xtrain1, Xtrain2))

      ytest = y[j:f]

      # turn ytrain into an array
      ytrain1 = y[0:j]
      ytrain2 = y[f:]
      ytrain = np.concatenate((ytrain1,ytrain2))

      # linear regression model fitting
      model = LinearRegression()
      model.fit(Xtrain.reshape(len(Xtrain),1), ytrain)

      # r2 scores
      training_r2.append(model.score(Xtrain.reshape(len(Xtrain),1), ytrain))
      testing_r2.append(model.score(Xtest.reshape(len(Xtest),1), ytest))

      avg = np.mean(testing_r2)
      std = np.std(testing_r2)
    
  return avg, std

# **Block 3: Print the average and standard deviation of the R^2 value for the different folds**

In [9]:
print('k=3:')
avg, std = kfold_score(data['duration'], data['waiting'], 3)   # unpack tuple that is returned from the function
print('average R^2:', avg)
print('std R^2:', std)
print('')

print('k=5:')
avg, std = kfold_score(data['duration'], data['waiting'], 5)
print('average R^2:', avg)
print('std R^2:', std)
print('')

print('k=10:')
avg, std = kfold_score(data['duration'], data['waiting'], 10)
print('average R^2:', avg)
print('std R^2:', std)
print('')

print('k=20:')
avg, std = kfold_score(data['duration'], data['waiting'], 20)
print('average R^2:', avg)
print('std R^2:', std)

k=3:
average R^2: 0.800452106299595
std R^2: 0.02982886623826577

k=5:
average R^2: 0.8064121141245482
std R^2: 0.021643210021753557

k=10:
average R^2: 0.7982771830742063
std R^2: 0.06730037541257397

k=20:
average R^2: 0.7779123591464386
std R^2: 0.14587623229874574


# **Block 4: In a text block, interpret this result**

The average R^2 value tells us how well the different partitions of X data (duration) predict the Y values (waiting). The 0.798 value when k=10 shows that in general, our model predicts the data pretty well even when going through different iterations of data. 

The standard deviation R^2 value tells us how much variance there is in our model. The .067 value when k=10 is pretty low, which shows that there is not much variance in our data, and that the splitting of the data doesn't affect our model.